daemons/controld/controld_control.c

```diff
@@ -66,17 +66,17 @@ do_ha_control(long long action,
 #endif
         }
 
-        if (registered == TRUE) {
+        if (registered) {
             controld_election_init(cluster->uname);
-            fsa_our_uname = cluster->uname;
+            controld_globals.our_nodename = cluster->uname;
```

Comment:
is it my imagination, or could this lead to a double free?

Reply:
I think with the way the code is structured now, the answer is no, because as far as I can tell nothing ever frees `cluster->uname`. `cluster` is function-static, only allocated once, and never freed. `pcmk__corosync_(dis)?connect()` doesn't free `cluster->uname` either.

However, it does look like there's a memory leak if we run the `A_HA_CONNECT` action more than once for a particular controller process. Each call to `pcmk__corosync_connect()` allocates a `cluster->uname` and `cluster->uuid`. It assumes they're currently NULL. I don't see any checks in the `do_ha_control()` call chain to ensure that if the `cluster` object has already been initialized, we either free its members or return early. Hmm...

Reply:
Ugh. I'm inclined to ignore it for this PR.

Reply:
I think A_HA_CONNECT is only ever used with I_STARTUP, which only happens once

Reply:
https://projects.clusterlabs.org/T598

Reply:
Thanks